terminate.go

```diff
@@ -0,0 +1,31 @@
+package main
+
+import (
+	"github.com/go-logr/logr"
+	"github.com/urfave/cli/v2"
+)
+
+const (
+	signalHandlerMetadataKeyName  = "signalHandler"
+	currentContextMetadataKeyName = "context"
+)
+
+func terminate(app *cli.App) {
+	log := app.Metadata[loggerMetadataKeyName].(logr.Logger)
+	if h, found := app.Metadata[signalHandlerMetadataKeyName]; found {
+		err := h.(func(context *cli.Context) error)(app.Metadata[currentContextMetadataKeyName].(*cli.Context))
+		if err != nil {
+			log.Error(err, "cannot properly shut down app")
+		}
+		return
+	}
+	log.Info("Shutting down")
+}
+
+// SetSignalHandler sets the given handler.
+// The handler will be invoked with given context in case of SIGINT or SIGTERM.
+// This setter should be called in an Action function (Before or Action itself) of a cli.App.
+func SetSignalHandler(ctx *cli.Context, handler func(context *cli.Context) error) {
```

Comment:
Why is this necessary? [`cli.Context`](https://pkg.go.dev/github.com/urfave/cli/v2#Context) already carries the cancelation context and thus cancelation signals if the app is run with [`App.RunContext`](https://pkg.go.dev/github.com/urfave/cli/v2#App.RunContext)?

I might be missing something?

Reply:
So, default App.Run has signal handling setup so that you can terminate the app. So far so good. However, if you want to execute additional code for a graceful shutdown, you have to use RunContext. With the `SetSignalHandler` I'm giving the subcommand the possibility to run custom shut down logic per subcommand.

Reply:
What I don't understand is why do we need this special signal handler function. The command gets a context that will be canceled when there's an interrupt. Can't the command can just listen and react on that? Something like this: 

```

func (c *exampleCommand) execute(context *cli.Context) error {
  select {
    case <-time.After(500 * time.Millisecond):
      fmt.Println("done")
    case <-ctx.Done():
      fmt.Println("halted")
    }
```

Reply:
Why not just do?

```go
func subcommand(ctx) {
	go func() {
		<-ctx.Done()
		_ = server.Shutdown()
	}()
	server.BlockingStart()
}
```

Reply:
Urgh, big :facepalm: I was so blind :smile: